### PR TITLE
Convert polls to Typescript / Immutable Records

### PR DESCRIPTION
**Pull request by renchap**
This is similar to other PRs:

- Add Typescript types for poll API JSON responses
- Create Immutable `Record` for the polls, to replace the existing `Immutable.Map` structure, which is not typed
- Create some helper methods to convert JSON from API to proper `Immutable` objects
- Rewrite the poll-related actions to Typescript, using `createAppAsyncThunk` (which will auto-generate all the related actions)
- Rewrite the polls reducer to Typescript, using what has been built above
- Update the (few) places with `state.polls` is accessed to use the new syntax (which preserves type information)

I think we can go one step further and transform the `Poll` objects to not use Immutable, as they are only accessed in one component and the reducers are very simple, but I preferred to keep this for a separate PR.

### CODE REVIEW

**Comment by renchap** [comment]
Typescript errors:

```
Error: app/javascript/mastodon/actions/polls.ts(40,19): error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type 'AsyncThunkAction<void, { poll: ApiPollJSON; }, { state: { announcements: any; dropdownMenu: DropdownMenuState; timelines: any; meta: Map<any, any>; alerts: any; loadingBar: any; ... 32 more ...; notificationRequests: any; }; ... 6 more ...; rejectedMeta?: unknown; }>' is not assignable to parameter of type 'AnyAction'.
Error: app/javascript/mastodon/actions/polls.ts(51,19): error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type 'AsyncThunkAction<void, { poll: ApiPollJSON; }, { state: { announcements: any; dropdownMenu: DropdownMenuState; timelines: any; meta: Map<any, any>; alerts: any; loadingBar: any; ... 32 more ...; notificationRequests: any; }; ... 6 more ...; rejectedMeta?: unknown; }>' is not assignable to parameter of type 'AnyAction'.
Error: Process completed with exit code 2.
```

I am not sure what causes this, if I comment out the `dispatch` override in the `createAsyncAppThunk` definition it goes away. Will ask some Redux people what we are doing wrong here.

File: app/javascript/mastodon/actions/polls.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  (args: { poll: ApiPollJSON }, { dispatch, getState }) => {
+    const { poll } = args;
+
+    dispatch(
+      importPolls({
+        polls: [createPollFromServerJSON(poll, getState().polls.get(poll.id))],
+      }),
+    );
+  },
+);
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/actions/polls.ts:23]: Afaik this does not need to be an action. But maybe it's just as simple like this.

File: app/javascript/mastodon/actions/polls.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+export const vote = createDataLoadingThunk(
+  'poll/vote',
+  ({ pollId, choices }: { pollId: string; choices: unknown }) =>
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/actions/polls.ts:27]: `unknown`? iirc it's a `string[]`

File: app/javascript/mastodon/api_types/polls.ts [hunk 3]
```diff
@@ -9,15 +9,15 @@ export interface ApiPollOptionJSON {
 
 export interface ApiPollJSON {
   id: string;
-  expires_at: string;
+  expires_at?: string;
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/api_types/polls.ts:12]: When is this undefined?

File: app/javascript/mastodon/api_types/polls.ts [hunk 4]
```diff
@@ -9,15 +9,15 @@ export interface ApiPollOptionJSON {
...
   expired: boolean;
   multiple: boolean;
   votes_count: number;
-  voters_count: number;
+  voters_count?: number;
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/api_types/polls.ts:16]: Same, when is it undefined?

**Comment by mjankowski** [comment]
This one is ~medium-ish old, needs rebase, has some unaddressed feedback ... is this in the "too much diff noise and fork damage for a lint-only change" and we should close? -- or is this still WIP/valid?

**Comment by ClearlyClaire** [comment]
That is definitely not a lint-only change, but it's also a bit outdated and needs more work…

**Review by Gargron — APPROVED** [review]

**Pull request merged**